### ISSUE TITLE
Parallel COPY may cause all placements to be marked inactive

### ISSUE THREAD

**Issue by marcocitus** [citusdata/citus#484]
This came up in the IRC channel. A user saw various shards with all placements in shardstate 3 after loading data using parallel COPY commands.

Because COPY for hash-partitioned tables makes a large number of connections to the workers (one to each placement), running it in parallel might cause max_connections to be exceeded. This results in  connection failures that cause placements to be marked inactive. Different COPY commands might see different placements fail, causing all placements to be marked inactive.

We could add a lock to allow only one COPY at a time. This would be a robust way to prevent problems resulting from the high number of connections made by COPY.  In general, we should prevent concurrent modifications of the same shard.

**Comment by anarazel** [comment]
Independent of locking considerations (which need work, although I don't think only one copy at a time is acceptable), I think we should consider adding some logic to handle different errors differently. E.g. it seems to make a lot of sense to handle ERRCODE_TOO_MANY_CONNECTIONS explicitly in COPY, for the reasons you name; I think we should just abort the whole COPY in that case, without marking individual shards as invalid.

**Comment by marcocitus** [comment]
That would be nice, though it doesn't seem like there's a way to detect this error code using libpq other than parsing the error message :-/. For some reason, the SQLSTATE can only be retrieved from a result, not from a connection.

**Comment by onderkalaci** [comment]
During the testing, I tried to reproduce this. But, when I run parallel COPY commands, I got an error instead of invalid placements.

The error message is below: 
```
DETAIL:  Remote message: could not fork new process for connection: Cannot allocate memory

could not fork new process for connection: Cannot allocate memory
could not fork new process for connection: Cannot allocate memory
```

**Comment by marcocitus** [comment]
It seems this is not caused by exceeding max_connections, but some other (OS?) limit.

**Comment by onderkalaci** [comment]
That could be the case because the worker instances were relatively small (i.e., c4.xlarge) on the cluster.
Basically,
Steps to reproduce this error: 
3 c4-xlarge workers
1 c4-2xlarge master
Distributed table with 256 shards and 2 replication.
Run, 10-15 concurrent COPY from different sessions

**Comment by marcocitus** [comment]
No longer relevant after #5381
